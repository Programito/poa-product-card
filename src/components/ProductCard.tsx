import { useProduct } from '../hooks/useProduct'
import React, { createContext } from 'react'

import styles from '../styles/styles.module.css'

import { ProductContextProps, Product, onChangeArgs, InitialValues, ProductCardHandlers } from '../interfaces/interfaces';
// import { ProductTitle } from './ProductTitle';
// import { ProductImage } from './ProductImage';
// import { ProductButtons } from './ProductButtons';



export const ProductContext = createContext({} as ProductContextProps)
const { Provider } = ProductContext;



// interface ProductButtonsProps {
//     increaseBy: (value: number)=> void,
//     counter: number
// }

// export const ProductButtons = ({counter, increaseBy}: ProductButtonsProps) => {


export interface Props {
    product: Product;
    // children?: ReactElement | ReactElement[],
    children: (args: ProductCardHandlers) => JSX.Element
    className?: string,
    style?: React.CSSProperties,
    onChange?: (args: onChangeArgs) => void;
    value?: number,
    initialValues?: InitialValues 
}

export const ProductCard = ({ children, product, className, style, onChange, value, initialValues }: Props) => {


    const { counter, increaseBy, maxCount, isMaxCountReached, reset } = useProduct({onChange, product, value, initialValues})

    return (
        <Provider value={{
            counter,
            increaseBy,
            maxCount,
            product,
        }}>
            <div 
                className={`${styles.productCard} ${className}`}
                style={style}
            >
                {/* <img className={styles.productImg} src='./coffee-mug.png' alt="Coffee Mug" /> */}
                {/* <img className="{styles.productImg}" src={noImage} alt="Coffee Mug" /> */}

                {children({
                    count: counter,
                    isMaxCountReached,
                    maxCount: initialValues?.maxCount,
                    product,
                    increaseBy,
                    reset
                })}

                {/* <ProductImage img={product.img} />
            <ProductTitle title={product.title} />

            <ProductButtons increaseBy={increaseBy} counter={counter} />  */}
            </div>
        </Provider>
    )
}


// ProductCard.Title = ProductTitle;
// ProductCard.Image = ProductImage;
// ProductCard.Buttons = ProductButtons;