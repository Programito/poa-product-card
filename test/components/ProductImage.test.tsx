import React from 'react';
import renderer from 'react-test-renderer';
import { ProductImage, ProductCard } from '../../src/components';

import { product2 } from '../data/product';

describe('ProductImage', () => {
  
    test('debe de mostrar el componente correctamente el productImage', () => {
        const wrapper = renderer.create(
            <ProductImage img="https://img.jpg" />
        )

        expect(wrapper.toJSON()).toMatchSnapshot();
    });

    test('debe de mostrar el componente con la imagen del producto', () => {
        const wrapper = renderer.create(
            <ProductCard product={product2}>
                {
                    () => (<ProductImage img="https://img.jpg" />)
                }
            </ProductCard>
            
        )

        expect(wrapper.toJSON()).toMatchSnapshot();
    });

});