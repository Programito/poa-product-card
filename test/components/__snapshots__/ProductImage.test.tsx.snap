// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage debe de mostrar el componente con la imagen del producto 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Product"
    className="productImg undefined"
    src="https://img.jpg"
  />
</div>
`;

exports[`ProductImage debe de mostrar el componente correctamente el productImage 1`] = `
<img
  alt="Product"
  className="productImg undefined"
  src="https://img.jpg"
/>
`;
